### ISSUE TITLE
Friction Model

### ISSUE THREAD

**Issue by bencannell** [tkoziara/parmec#13]
Hi @tkoziara,

We would like to include functionality to model friction contact within parmec.

What we are trying to achieve is to model some lateral shear resistance between points on particles, which is proportional to the load experienced by a spring which connects these particles. For example, we have springs which model direct vertical contact between vertically stacked components, we would like to model planar friction behaviour between these components. The normal to this plane would be defined by the associated direct contact spring.

Do you think it would be possible to add this into the SPRING definition? Or perhaps a separate command?

Perhaps @stevebrasier, could add some extra clarity to this request?

Best regards,

Ben

**Comment by tkoziara** [comment]
Most naturally, an optional parameter *friction* could be added to the SPRING command to accompany the contact plane spring model or a two-point-spring with constant direction. A ratio ks/kn could be used to inherit the tangential spring stiffness from the normal one; similarly, the normal damper curve could be reused in the tangential direction (same ratio could be applied); the normal direction would be the constant direction or the plane normal; by default *friction = 0* and defining *friction > 0* would enable this; history of local tangential displacements could be tracked and thus stick-slip behaviour could be modelled; Example interface:

SPRING(... , friction, kskn)

Would that be ok?

**Comment by bencannell** [comment]
Hi @tkoziara,

Yes I think your suggested interface would be appropriate.

Many thanks,

Ben

**Comment by tkoziara** [comment]
This is now initially implemented and tested; Please see [this blog post](http://parmes.org/blog/parmec_spring_contact.html); closing issue for now.

**Comment by bencannell** [comment]
Thank you @tkoziara, that was quick! I'll test this week

**Comment by bencannell** [comment]
Hi @tkoziara,

I've tested the friction definition and I think I may have mislead you regarding the ks/kn definition. I would not expect there to be any tangential stiffness/damping inherited from the normal spring definition, rather just a single tangential force which is the normal spring force * friction coefficient. This force would be applied to the component of the particle's velocity tangential to the normal force (i.e. tangential to the spring direction), and would "effectively" become zero once the velocity of the particle becomes zero.

I've attached a test case which models a particle dropping onto a plane under gravitational loading with an initial tangential velocity. The particle decelerates due to a constant friction force once the particle has "settled" vertically. However, the particle unexpectedly experiences a negative tangential acceleration resulting in a negative tangential velocity, which does not appear physical. I would expect the effective friction force to be zero once the velocity becomes zero and hence there is no further acceleration of the particle (i.e. the particle will have "stuck" to the plane, so although there is a static friction force the velocity vector is 0).

test case:
[test-friction.txt](https://github.com/tkoziara/parmec/files/2054781/test-friction.txt)

velocity plot:
![vel-friction](https://user-images.githubusercontent.com/29304199/40733778-6ad2f5fa-642e-11e8-894b-bd019b7ca148.png)

We would not expect to need to apply any tangential damping to the model to negate this behaviour. This will become complicated with the way we currently model "springs" and "dampers" since these are defined using separate SPRING definitions (e.g. see the test-friction.txt).

What do you think? Is it possible to change the SPRING definition such that stick/slip behaviour is modelled without defining tangential spring/damper values (ks/kn)?

Best regards,

Ben

**Comment by tkoziara** [comment]
Let me have a look at this example and get back; A simpler friction model is of course possible: one that applies friction always opposing slip velocity; In this case there is no need for tangential springs and dampers - but also there is now way to model frictional sticking and stick-slip transition; perhaps we can have a way of including both approaches and switching between them:) I'll have a look.

**Comment by tkoziara** [comment]
Ben - I have modified the model now so that **kskn** = 0 is allowed in the SPRING command (see [updated docs](http://parmes.org/parmec/input_commands.html#spring)). Practically the same effect can be obtained either by using *kskn = 0* or by using *kskn = very large number*. See figure below.
![kskncompare](https://user-images.githubusercontent.com/2198068/40842598-aa985fc8-65ae-11e8-9b18-a7f0b306bc79.jpg)
There is an oscillation though when the particle comes to rest. When I print out friction force components it keeps switching between positive and negative every time step. You can see the pattern in the velocity here:
![vel-friction](https://user-images.githubusercontent.com/2198068/40842932-adb55746-65af-11e8-9523-ae515cab816f.png)
*Why did you use two springs in the example? (the contact is undamped)* When enabling damping in the contact spring and disabling the other spring I am getting awkward results:) So perhaps it is a good example for testing this. Thank you. Please, feel free to investigate from your end. I will also continue looking and let know when there is more. Do use the latest commit to test the above.

**Comment by bencannell** [comment]
Hi @tkoziara,

Thank you for looking into this and providing the figures. The behavior now looks more similar to what we expect.

I included two vertical springs in the model since this is how we model "contact" interactions in our large array models. One of the SPRING commands defines just a spring curve (the damping curve is set to [0.0, 0.0,...]), we then use a second SPRING command to represent a damper. I think this gives us more control over how the damping is applied and it is a direct translation of how the spring/damper pairs are modelled in the LS-DYNA models.

I'll let you know how I get on with the updated source code.

Ben

**Comment by bencannell** [comment]
Hi @tkoziara,

Just an initial observation, but the magnitude of the high frequency oscillations between +ve and -ve velocity appear to be very timestep dependent:

timestep = 0.5 * CRITICAL()

![vel-friction0 5crit](https://user-images.githubusercontent.com/29304199/40845969-2166ef24-65b0-11e8-9cb0-0f8910230e7a.png)


timestep = 0.1 * CRITICAL()

![vel-friction0 1crit](https://user-images.githubusercontent.com/29304199/40845974-283b2874-65b0-11e8-93e5-ea99616f7e96.png)

**Comment by tkoziara** [comment]
Yes, indeed, I also noticed that; Basically the friction force keeps switching from -f_tiny to +f_tiny while the magnitude of it is time-step dependent; So this seems to be a numerical artefact of some sort; I'll try to understand and eliminate it; I will look into creating a one spring model which reproduces that - if you were able to construct another one and paste it in - that might be helpful:)

**Comment by tkoziara** [comment]
Oh, look, I fixed a little bug (e0cdc1f), and with this example:
[single-contact.py.gz](https://github.com/tkoziara/parmec/files/2067316/single-contact.py.gz)
we can get these figures:
![single-contact-displacement](https://user-images.githubusercontent.com/2198068/40904972-be870b4e-67dc-11e8-8618-0c19a844d69d.png)
![single-contact-force](https://user-images.githubusercontent.com/2198068/40904973-bea3632a-67dc-11e8-90db-2df852bd8f43.png)
![single-contact-velocity](https://user-images.githubusercontent.com/2198068/40904974-bed63624-67dc-11e8-9e32-70c97b04f233.png)
So the issue is now isolated with a simplest possible test:) I am wondering whether it will be possible to "fix" it or whether it is a modelling error pertinent to this type of modelling. I will share more when some conclusions come; (by the way, below I accidentally closed and reopened the issue)

**Comment by tkoziara** [comment]
Ben; have a look below; the same model as above, this time using *kskn* - in this case we are able to model a transition from slip to stick. You see - when the time steps on both springs are set to coincide there is a way to bring the tangential motion to stop; I am thinking that perhaps with the simpler model (*kskn = 0*, which is numerically the same as *kskn = big*) we may not be able to model tangential sticking (hence coming to rest) except approximately - using a sufficiently small step; I'll have a little more though about this - but please have a look and let me know about how you feel about this:)
![kskncompare2](https://user-images.githubusercontent.com/2198068/40905916-a14b73a0-67df-11e8-9e10-ba77dcc0578c.jpg)

**Comment by bencannell** [comment]
Hi @tkoziara,

@stevebrasier and I have been discussing this and have the following thoughts:

- We think the reason you've included a “tangential” spring/damper is essentially so that the direction for the frictional force is simple to define. However this means there are forces proportional to tangential displacement and tangential velocities - this seems non-physical.
- With us using kskn=0 with the current implementation, we think we’re effectively saying that there’s no static friction, only dynamic. We didn’t really mean that, we were just trying to get rid of the above non-physical forces.
- We wonder whether the direction of the frictional force could be calculated from:
o	The relative velocity in the tangential plane, when this is non-zero
o	The projection of the net force on the particle onto the tangential plane, when the above is zero.
(with the magnitude of the frictional force obviously calculated from the “normal” spring+damper force x mu)

What do you think?

Best regards,

Ben

**Comment by tkoziara** [comment]
Hi,

*(1st bullet)* Using tangential springs and dampers is one of the several common approaches for modelling friction in explicit rigid body frameworks (e.g. in DEM). Another (less) common option is to use "viscous damping", which is what we have with kskn = 0. Have a look at [these slides](https://pharmahub.org/resources/254/download/psl_purdue_dem_lecture09_tangentialcontactforcemodels.pdf) for example - you can tell me which model (give page number) from those slides you would like to have. *(I will also have a look at those slides and perhaps improve some aspects of the current implementation; if so I will let know in this thread)*

*(2nd bullet)* There would be no way of modelling static friction without those "non-physical" forces in a purely rigid, explicit modelling framework. Either this or NSCD-like implicit approaches (which turn parmec into solfec) - are the available options.

*(3rd bullet)* When kskn = 0, the direction of the frictional force is opposite to the relative tangential velocity. When kskn > 0, in the current implementation, this will depend on the slip history and the force will oppose the tangential displacement. But, the slip displacement history is truncated at the that sufficient to reach the friction cone (so that you can model a large slip displacement with "a short slip memory" which seems preferable - otherwise such large slip displacement would unrealistically influence the direction of the friction force).

What do you think?

:)
Tomek
